ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala

```diff
@@ -0,0 +1,238 @@
+        if (contractStateEventsBatch.nonEmpty) {
+          participantInMemoryState.contractStateCaches.push(contractStateEventsBatch)
+        }
+
+      case _: TransactionLogUpdate.LedgerEndMarker =>
+        // TODO LLP: Remove TransactionLogUpdate.LedgerEndMarker
+        ()
+    }
+
+  private def updateLedgerEnd(
+      participantInMemoryState: ParticipantInMemoryState
+  )(lastOffset: Offset, lastEventSequentialId: Long)(implicit
+      loggingContext: LoggingContext
+  ): Unit = {
+    participantInMemoryState.ledgerEndCache.set((lastOffset, lastEventSequentialId))
+    // the order here is very important: first we need to make data available for point-wise lookups
+    // and SQL queries, and only then we can make it available on the streams.
+    // (consider example: completion arrived on a stream, but the transaction cannot be looked up)
+    participantInMemoryState.dispatcher().signalNewHead(lastOffset)
+    logger.debug(s"Updated ledger end at offset $lastOffset - $lastEventSequentialId")
+  }
+
+  private def convertToContractStateEvents(
+      tx: TransactionLogUpdate
+  ): Vector[ContractStateEvent] =
+    tx match {
+      case tx: TransactionLogUpdate.Transaction =>
+        tx.events.iterator.collect {
+          case createdEvent: TransactionLogUpdate.CreatedEvent =>
+            ContractStateEvent.Created(
+              contractId = createdEvent.contractId,
+              contract = Contract(
+                template = createdEvent.templateId,
+                arg = createdEvent.createArgument,
+                agreementText = createdEvent.createAgreementText.getOrElse(""),
+              ),
+              globalKey = createdEvent.contractKey.map(k =>
+                Key.assertBuild(createdEvent.templateId, k.unversioned)
+              ),
+              ledgerEffectiveTime = createdEvent.ledgerEffectiveTime,
+              stakeholders = createdEvent.flatEventWitnesses.map(Party.assertFromString),
+              eventOffset = createdEvent.eventOffset,
+              eventSequentialId = createdEvent.eventSequentialId,
+            )
+          case exercisedEvent: TransactionLogUpdate.ExercisedEvent if exercisedEvent.consuming =>
+            ContractStateEvent.Archived(
+              contractId = exercisedEvent.contractId,
+              globalKey = exercisedEvent.contractKey.map(k =>
+                Key.assertBuild(exercisedEvent.templateId, k.unversioned)
+              ),
+              stakeholders = exercisedEvent.flatEventWitnesses.map(Party.assertFromString),
+              eventOffset = exercisedEvent.eventOffset,
+              eventSequentialId = exercisedEvent.eventSequentialId,
+            )
+        }.toVector
+      case _ => Vector.empty
+    }
+
+  private def updateToTransactionAccepted(
+      offset: Offset,
+      u: TransactionAccepted,
+  ): TransactionLogUpdate.Transaction = {
+    val rawEvents = u.transaction.transaction
```

Comment:
is this a copy paste from respective code in UpdateToDbDto? Shall we extract into some utility function? Maybe on LF side?

Reply:
Good catch, thanks. Will do so!